internal/test/environment/kubernetes/setup.go

```diff
@@ -0,0 +1,352 @@
+			hostCfg: &container.HostConfig{RestartPolicy: container.RestartPolicy{Name: "always"}},
+		},
+		{
+			name:    "kube-apiserver",
+			cfg:     apiServerContainerConfig(hyperkubeImage, ""),
+			hostCfg: apiServerHostConfig,
+		},
+		{
+			name:    "kube-controller-manager",
+			cfg:     controllerManagerContainerConfig(hyperkubeImage),
+			hostCfg: controllerManagerHostConfig,
+		},
+		{
+			name:    "kube-scheduler",
+			cfg:     schedulerContainerConfig(hyperkubeImage),
+			hostCfg: schedulerHostConfig,
+		},
+	}); err != nil {
+		return err
+	}
+
+	fmt.Fprintln(w, "Write kubeconfig file...")
+	if err := writeKubeConfig(hostname, kubeconfig, ca); err != nil {
+		return err
+	}
+
+	fmt.Fprintln(w, "Starting kube-dns...")
+	err = withRetries(func() error { return applyYaml(ctx, kubeconfig, kubeDNSYaml) }, 10, 3*time.Second)
+	if err != nil {
+		return err
+	}
+
+	fmt.Fprintln(w, "Starting weave-net (CNI plugin)...")
+	err = withRetries(func() error { return applyYaml(ctx, kubeconfig, cniYaml) }, 10, 3*time.Second)
+	if err != nil {
+		return err
+	}
+
+	return starter.startContainer("kube-proxy", proxyContainerConfig(hyperkubeImage), proxyHostConfig)
+}
+
+func writeKubeConfig(hostname, kubeconfig string, ca *certAuthority) error {
+	kubecfg, err := makeKubeConfig(kubeConfigSpec{
+		clientName:    "kubernetes-admin",
+		organization:  []string{"system:masters"},
+		apiserverHost: hostname,
+	}, ca)
+	if err != nil {
+		return err
+	}
+	return ioutil.WriteFile(kubeconfig, []byte(kubecfg), 0644)
+}
+
+func createKubeConfigVolume(ctx context.Context, c client.APIClient, hostname string, ca *certAuthority, certs map[string]cert) error {
+	configSpecs := map[string]kubeConfigSpec{
+		"admin.config": {
+			clientName:    "kubernetes-admin",
+			organization:  []string{"system:masters"},
+			apiserverHost: "kube-apiserver",
+		},
+		"kubelet.config": {
+			clientName:   "system:node:" + hostname,
+			organization: []string{"system:nodes"},
+		},
+		"proxy.config": {
+			clientName:   "kube-proxy",
+			organization: []string{"system:node-proxier"},
+		},
+		"controller-manager.config": {
+			clientName:    "system:kube-controller-manager",
+			apiserverHost: "kube-apiserver",
+		},
+		"scheduler.config": {
+			clientName:    "system:kube-scheduler",
+			apiserverHost: "kube-apiserver",
+		},
+	}
+	if _, err := c.VolumeCreate(ctx, volumetypes.VolumesCreateBody{Name: "kube-config"}); err != nil {
+		return err
+	}
+	cont, err := c.ContainerCreate(ctx, &container.Config{
+		Image: busyboxImage,
+		Cmd:   []string{"top"},
+	}, &container.HostConfig{
+		Mounts: []mount.Mount{
+			{
+				Type:   mount.TypeVolume,
+				Source: "kube-config",
+				Target: "/kube-config",
+			},
+		},
+	}, nil, "kube-config-init")
+	if err != nil {
+		return err
+	}
+	defer c.ContainerRemove(ctx, cont.ID, types.ContainerRemoveOptions{Force: true})
```

Comment:
defer the remove after the start?

Reply:
Not really, as soon as the container is created I want to make sure I remove it anyway 👼